File: 2026-01/src/day_31/boj_16953/Main.cpp
#include <bits/stdc++.h>
using namespace std;

int bfs(int a, int b) {
    queue<int> q;
    q.push(a);

    set<int> vis;
    vis.insert(a);

    int dist = 1;

    while (!q.empty()) {
        int sz = q.size();

        while (sz--) {
            int cur = q.front();
            q.pop();

            if (cur == b) return dist;

            if (cur * 2LL <= b && !vis.count(cur * 2)) {
                q.push(cur * 2);
                vis.insert(cur * 2);
            }

            if (cur * 10LL + 1 <= b && !vis.count(cur * 10 + 1)) {
                q.push(cur * 10 + 1);
                vis.insert(cur * 10 + 1);
            }
        }

        dist++;
    }

    return -1;
}

int main() {
    ios::sync_with_stdio(0);
    cin.tie(0);

    int a, b;
    cin >> a >> b;
    cout << bfs(a, b);
}

File: 2026-01/src/day_31/boj_16953/Main.java
package day_31.boj_16953;

import java.io.*;
import java.util.*;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());

        int a = Integer.parseInt(st.nextToken());
        int b = Integer.parseInt(st.nextToken());

        System.out.println(bfs(a, b));
    }

    static int bfs(int a, int b) {
        Queue<Integer> q = new ArrayDeque<>();
        q.offer(a);

        Set<Integer> vis = new HashSet<>();
        vis.add(a);

        int dist = 1;

        while (!q.isEmpty()) {
            int sz = q.size();

            while (sz-- > 0) {
                int cur = q.poll();
                if (cur == b) return dist;

                if (cur * 2L <= b && !vis.contains(cur * 2)) {
                    q.offer(cur * 2);
                    vis.add(cur * 2);
                }

                if (cur * 10L + 1 <= b && !vis.contains(cur * 10 + 1)) {
                    q.offer(cur * 10 + 1);
                    vis.add(cur * 10 + 1);
                }
            }

            dist++;
        }

        return -1;
    }
}
